vmf/scripts/mods/vmf/modules/vmf_package_manager.lua

```diff
@@ -13,166 +71,139 @@ local _loaded_packages = {}
 -- ##### VMF internal functions and variables ##########################################################################
 -- #####################################################################################################################
 
--- Loads queued packages one at a time
+-- Loads queued packages one at a time.
 function vmf.update_package_manager()
-  local loading_package = _loading_package
-  if loading_package and loading_package.resource_package:has_loaded() then
-    loading_package.resource_package:flush()
-
-    _loaded_packages[loading_package.mod][loading_package.package_name] = loading_package.resource_package
-    _loading_package = nil
-
-    -- The callback has to be called last, so that any calls to `has_package_loaded` or `is_package_loading`
-    -- return the correct value
-    vmf.safe_call_nr(loading_package.mod, {"'%s' package loaded callback", loading_package.package_name},
-                      loading_package.callback, loading_package.package_name)
-  end
-
-  local queued_package = _queued_packages[1]
-  if queued_package and not _loading_package then
-    _loading_package = queued_package
-    table.remove(_queued_packages, 1)
+  local queued_package_name = _queued_packages[1]
+  if queued_package_name then
+    local package_data = _packages[queued_package_name]
+    if package_data.status == "loading" and package_data.resource_package:has_loaded() then
+      flush_package(queued_package_name)
+      table.remove(_queued_packages, 1)
+    end
 
-    _loading_package.resource_package:load()
+    if package_data.status == "queued" then
+      package_data.resource_package:load()
+      package_data.status = "loading"
+    end
   end
 end
 
 
--- Forcefully unloads all mods and cleans the queue.
+-- Forcefully unloads all not unloaded packages.
 function vmf.unload_all_resource_packages()
-  for mod, packages in pairs(_loaded_packages) do
-    for package_name in pairs(packages) do
-      mod:warning(
-        "Force-unloading package '%s'. Please make sure to properly release packages when the mod is unloaded",
-        package_name
-      )
-      mod:unload_package(package_name)
-    end
-  end
-
-  _queued_packages = {}
+  for package_name, package_data in pairs(_packages) do
+    local package_status = package_data.status
 
-  if _loading_package then
-    _loading_package.mod:warning(
-      "Still loading package '%s'. Memory leaks may occur when unloading while a package is loading.",
-      _loading_package.package_name
-    )
+    if package_status == "loaded" then
+      package_data.mod:warning(WARNINGS.force_unloading_package, package_name)
+      package_data.mod:unload_package(package_name)
+    end
 
-    _loading_package.callback = function(package_name)
-      _loading_package.mod:unload_package(package_name)
+    if package_status == "loading" then
+      package_data.mod:warning(WARNINGS.still_loading_package, package_name)
```

Comment:
Packages which are currently `"loading"` will not be released, potentially causing a leak. I would *either*:

+ Increase the message severity to `error` to reflect that the solicited action `unload_all_resource_packages` could not be fully carried out, or
+ Restore SirAiedail's version that replaced the callback with one that unloads the package:

```suggestion
      _loading_package.callback = callback(_loading_package.mod, "unload_package")
```

Reply:
This can happen only when reloading VMF (i.e. only during development) and you can't really do anything about it. Also, it shouldn't break the game (I suppose). So we can just warn modder about this incident, since it's not in their power to fix it. If you still think `error` is necessary, I can change that.

The callback won't be called because VMF will be disabled by this moment.